ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala

```diff
@@ -963,6 +966,93 @@ private class PostgresLedgerDao(
     }
   }
 
+  // Note: package upload is idempotent
+  private val SQL_INSERT_PACKAGE =
+    """insert into packages(package_id, submission_id, source_description, size, known_since, ledger_offset, package)
+          |select {package_id}, {submission_id}, {source_description}, {size}, {known_since}, ledger_end, {package}
+          |from parameters
+          |on conflict (package_id)
+          |do nothing
+          |""".stripMargin
+
+  private val SQL_SELECT_PACKAGES =
+    SQL("""select package_id, source_description, known_since, size
+          |from packages
+          |""".stripMargin)
+
+  private val SQL_SELECT_PACKAGE =
+    SQL("""select package
+          |from packages
+          |where package_id = {package_id}
+          |""".stripMargin)
+
+  case class ParsedPackageData(
+      packageId: String,
+      sourceDescription: Option[String],
+      size: Long,
+      knownSince: Date)
+
+  private val PackageDataParser: RowParser[ParsedPackageData] =
+    Macro.parser[ParsedPackageData](
+      "package_id",
+      "source_description",
+      "size",
+      "known_since"
+    )
+
+  override def listLfPackages: Future[Map[PackageId, PackageDetails]] =
+    dbDispatcher.executeSql { implicit conn =>
+      SQL_SELECT_PACKAGES
+        .as(PackageDataParser.*)
+        .map(
+          d =>
+            PackageId.assertFromString(d.packageId) -> PackageDetails(
+              d.size,
+              d.knownSince.toInstant,
+              d.sourceDescription))
+        .toMap
+    }
+
+  override def getLfArchive(packageId: PackageId): Future[Option[Archive]] =
+    dbDispatcher.executeSql { implicit conn =>
+      SQL_SELECT_PACKAGE
+        .on(
+          "package_id" -> packageId
+        )
+        .as[Option[Array[Byte]]](SqlParser.byteArray("package").singleOpt)
+        .map(data => Archive.parseFrom(Decode.damlLfCodedInputStreamFromBytes(data)))
+    }
+
+  override def uploadLfPackages(
+      packages: List[(Archive, PackageDetails)]): Future[UploadPackagesResult] = {
+    dbDispatcher.executeSql { implicit conn =>
+      Try {
+        val submissionId = UUID.randomUUID().toString
```

Comment:
nitpick: in general you want to do as little as possible within `dbDispatcher.executeSql` to avoid occupying the SQL dedicated threads with CPU heavy stuff. As a rule of thumb, only code requiring a connection should be within such blocks.

Reply:
Makes sense, I will move the task of creating the submission ID to `SqlLedger`.